### PR TITLE
Fix IndexOf/LastIndexOf behavior for float/double types to match Java semantics, #146

### PR DESCRIPTION
**Pull request by paulirwin**
- Add IndexOf<T> wrapper methods to Arrays class with special handling for float, double, and nullable variants
- Add LastIndexOf<T> wrapper methods to Arrays class with special handling for float, double, and nullable variants
- Update List<T>.IndexOf() and LastIndexOf() to use new Arrays methods instead of Array methods
- Extract duplicated search logic into helper methods IndexOfFloatingPoint<T> and LastIndexOfFloatingPoint<T>
- Add comprehensive test cases for NaN and signed zero behavior

Fixes issue where NaN != NaN and +0.0 == -0.0 by leveraging J2N's EqualityComparer<T> which implements Java semantics.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/J2N/Collections/Arrays.cs [hunk 1]
```diff
@@ -390,5 +392,224 @@ private static class EmptyArrayHolder<T>
...
+                ThrowHelper.ThrowArgumentOutOfRange_IndexMustBeLessOrEqualException(startIndex);
+            if (count < 0 || startIndex > array.Length - count)
+                ThrowHelper.ThrowCountArgumentOutOfRange_ArgumentOutOfRange_Count(count);
+
+            // Special handling for float and double types to match Java's behavior
+            if (typeof(T) == typeof(float))
+            {
+                float floatValue = Unsafe.As<T, float>(ref value);
+                float[] floatArray = Unsafe.As<T[], float[]>(ref array);
+                return IndexOfFloatingPoint(floatArray, floatValue, startIndex, count, JCG.EqualityComparer<float>.Default);
```
Comment (NightOwl888) [inline src/J2N/Collections/Arrays.cs]: I probably would have went with using `ref T` instead of `T[]` here and brought over the support from [`SpanHelpers`](https://github.com/dotnet/runtime/blob/v10.0.0-rc.2.25502.107/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs#L1480), but being that this is such an edge case that we will probably never encounter it, I think this is fine.

**Review by NightOwl888 — CHANGES_REQUESTED** [review]

File: tests/Xunit/J2N.Collections.Generic.List.Tests/List.Generic.Tests.IndexOf.cs [hunk 2]
```diff
@@ -252,5 +252,292 @@ public void LastIndexOf_int_int_OutOfRangeExceptions(int count)
...
+        [Fact]
+        public void LastIndexOf_FloatNaN()
+        {
+            if (typeof(T) != typeof(float))
+                return;
+
+            // Test with different NaN bit patterns
+            float nanDefault = float.NaN; // 0x7FC00000
+            float nanVariant1 = BitConverter.Int32BitsToSingle(0x7FC00001);
+            float nanVariant2 = BitConverter.Int32BitsToSingle(0xFFC00001);
```
Comment (NightOwl888) [inline tests/Xunit/J2N.Collections.Generic.List.Tests/List.Generic.Tests.IndexOf.cs]: The tests are now failing due to the fact that `BitConverter` didn't have this overload on .NET Framework. Ideally, these would be `const float` values rather than converted on the fly so we don't have to rely on a conversion to happen.

But failing that, we should switch these to use `J2N.BitConversion` rather than `System.BitConverter`. Note that the rest of J2N doesn't use `BitConverter` at all for floating point conversions outside of the BCL tests, both to align with the JDK and to cover this missing overload.

File: src/J2N/Collections/Generic/EqualityComparer.cs [hunk 3]
```diff
@@ -206,6 +206,10 @@ public override bool Equals(double x, double y)
             if (double.IsNaN(x) && double.IsNaN(y))
                 return true;
 
+            // If only one is NaN, they're not equal
```
Comment (NightOwl888) [inline src/J2N/Collections/Generic/EqualityComparer.cs:209]: It looks like this implementation may have predated the `BitConversion` class and was just an adaption of the `compare()` method in Apache Harmony. So, we ended up with a bloated implementation. I opened #183 so we can address this and align with Harmony.

File: src/J2N/Collections/Generic/EqualityComparer.cs [hunk 4]
```diff
@@ -285,6 +289,10 @@ public override bool Equals(float x, float y)
             if (float.IsNaN(x) && float.IsNaN(y))
                 return true;
 
+            // If only one is NaN, they're not equal
```
Comment (NightOwl888) [inline src/J2N/Collections/Generic/EqualityComparer.cs:292]: See my comment above on the `Equals(double, double)` overload.

**Comment by paulirwin** [comment]
@NightOwl888 reminder of this PR. I addressed the PR feedback a while back.

**Review by NightOwl888 — COMMENTED** [review]
During my first review of this, I was thinking that it probably didn't matter that we were going to have expensive `IndexOf` lookups for floating point types. But those assumptions were based on:

- It isn't very useful to make `float` or `double` a key to a hash table
- `IndexOf()` and `LastIndexOf()` would be the only surface area affected

Unfortunately, both of those assumptions were incorrect. In fact, it can be very useful to put `float` and `double` values into a list and this has nothing to do with hash tables. Since the `Compare()` and `Remove()` methods rely on `IndexOf()` lookups, this will cause a significant difference in performance from the BCL, which uses SIMD optimizations and raw byte comparisons for this operation.

So, the goal became:

- Don't slow down all `float` and `double` IndexOf operations, just account for the special cases
- Don't impact performance at all for `int`, `long`, or any other types the patch doesn't apply to
- Put the implementation in the `CollectionUtil` class, because this is applicable to Java collections, not Java arrays (this one was on me)
- Document the implementation appropriately because this is a very narrow use case, not meant for general use

I was also wrong about NaN - .NET already covers that match correctly in the `Arrays.IndexOf()` implementation.

It seemed like a lot to ask of you to start over from scratch with a corrected set of requirements, so I went ahead and did the work myself in #182. I was able to use the tests here to ensure there were no gaps in our assumptions and your edits to the `DoubleComparer.Equals()` and `SingleComparer.Equals()` methods prompted me to check the Apache Harmony implementation, which means we can simplify (see #183). So, not all of your work on this was in vain.

I am closing this as superseded by #182.
